compiler/passes/insertWideReferences.cpp

```diff
@@ -45,11 +229,268 @@ static void insertWideCastTemps();
 static void derefWideStringActuals();
 static void derefWideRefsToWideClasses();
 static void widenGetPrivClass();
+static void moveAddressSourcesToTemp();
+static void fixAST();
 
-AggregateType* wideStringType = NULL;
+//
+// Miscellaneous utility functions to help manage the AST
+//
 
-static void
-buildWideClass(Type* type) {
+static bool isFullyWide(BaseAST* bs) {
+  return bs->typeInfo()->symbol->hasEitherFlag(FLAG_WIDE_CLASS, FLAG_WIDE_REF);
+}
+
+static bool valIsWideClass(BaseAST* bs) {
+  return bs->getValType()->symbol->hasFlag(FLAG_WIDE_CLASS);
+}
+
+static Type* getNarrowType(BaseAST* bs) {
+  //
+  // TODO: This should really return a completely narrow ref, and not
+  // a ref to a wide class.
+  //
+  if (isFullyWide(bs)) {
+    return bs->typeInfo()->getField("addr")->typeInfo();
+  }
+  else {
+    return bs->typeInfo();
+  }
+}
+
+static bool hasSomeWideness(Symbol* sym) {
+  return isFullyWide(sym) || valIsWideClass(sym);
+}
+
+static bool isRef(BaseAST* bs) {
+  return bs->typeInfo()->symbol->hasEitherFlag(FLAG_REF, FLAG_WIDE_REF);
+}
+
+static bool isObj(BaseAST* bs) {
+  return (isClass(bs->typeInfo()) && !isRef(bs)) || bs->typeInfo()->symbol->hasFlag(FLAG_WIDE_CLASS);
+}
+
+static bool typeCanBeWide(Symbol *sym) {
+  TypeSymbol* ts = sym->type->symbol;
+  bool bad = sym->hasFlag(FLAG_EXTERN) ||
```

Comment:
I think we will want to support wide references to extern types, but that is beyond the scope of this pass.  Just a heads-up: the special treatment of extern types may be removed in follow-on AMM work.